packages/flutter/lib/src/widgets/context_menu_controller.dart

```diff
@@ -0,0 +1,114 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'basic.dart';
+import 'context_menu_button_item.dart';
+import 'editable_text.dart';
+import 'framework.dart';
+import 'inherited_theme.dart';
+import 'navigator.dart';
+import 'overlay.dart';
+
+/// Builds a context menu at [primaryAnchor] if possible, otherwise at
+/// [secondaryAnchor].
+typedef ContextMenuBuilder = Widget Function(
+  BuildContext context,
+  Offset primaryAnchor,
+  [Offset? secondaryAnchor]
+);
+
+/// A builder function that builds a context menu given a list of
+/// [ContextMenuButtonItem]s representing its children.
+///
+/// See also:
+///
+///   * [EditableTextContextMenuButtonItemsBuilder], which receives this as a
+///     parameter.
+typedef ToolbarButtonWidgetBuilder = Widget Function(
+  BuildContext context,
+  List<ContextMenuButtonItem> buttonItems,
+);
+
+/// A function that builds a widget to use as the text selection toolbar for
+/// [EditableText].
+///
+/// See also:
+///
+///  * [ContextMenuBuilder], which is the generic type for any context menu
+///    builder, not just for the editable text selection toolbar.
+///  * [SelectableRegionToolbarBuilder], which is the builder for
+///    [SelectableRegion].
+typedef EditableTextToolbarBuilder = Widget Function(
+  BuildContext,
+  EditableTextState,
+  Offset,
+  [Offset?]
+);
+
+/// Builds and manages a context menu at a given location.
+///
+/// There can only ever be one context menu shown at a given time.
+///
+/// {@tool dartpad}
+/// This example shows how to use a GestureDetector to show a context menu
+//  anywhere in a widget subtree that receives a right click or long press.
+///
+/// ** See code in examples/api/lib/material/context_menu/anywhere.dart **
+/// {@end-tool}
+class ContextMenuController {
+  ContextMenuController._();
+
+  // The OverlayEntry is static because only one context menu can be displayed
```

Comment:
Good question. I hadn't thought about it, but it should work out correctly as-is.  There can be only one context menu in any window of a Flutter app, which is correct.

Reply:
I've refactored ContextMenuController to not be a singleton, and now you can just do `controller.isShown`.  See https://github.com/flutter/flutter/pull/107193#issuecomment-1241333434

Reply:
Can we give each menu a focus node instead of having a global registry? Then each menu will be notified when it's no longer in the path of the primary focus and can optionally dismiss itself. Not sure if how this applies to SeletableRegion, but SelectableRegion needs a focus node to do shortcuts anyways?

Reply:
@LongCatIsLooong Do you mean to get rid of the global removeAny method?  Let me know if this is what you mean: https://github.com/justinmc/flutter/commit/a483fbb9fdf9d7b23668fe52c802af019a4bd4ef

I didn't need to touch SelectableRegion because it uses `.remove` directly.  Maybe it's a bit strange using a FocusNode to remove the menu in my context_menu_controller.0 example.  Not sure if this is exactly what you mean though.

Reply:
In my examples I'm using removeAny to close the menu that's managed by the TextField, see: https://github.com/flutter/flutter/pull/107193/files#diff-c37e45d149cf86a6565f1eb32f738d5faaf4170b8bd4f1f6421651f0af2f1bfaR56

Without that I might need to pass the instance of ContextMenuController via the contextMenuBuilder parameters.  Do you think that's a better option?

I think that the owner of a context menu should expect that their menu can be closed at any time.  If the users interacts with any other part of the app then the menu will be closed.  The owner can listen to `onRemove` if they want.

Maybe I should make it more clear in the docs that the menu can be closed from anywhere?